### ISSUE TITLE
CLI options to follow/monitor deployments to completion

### ISSUE THREAD

**Issue by schmichael** [hashicorp/nomad#6818]
### Nomad version

Nomad 0.10.2

### Issue

`nomad deployment status` lacks to ability to block until a deployment is complete similar to log following/tailing or drain monitoring.

`nomad job run` monitors by default but not until a deployment is complete.

### Proposal

Add `-monitor` command to `nomad deployment status` that blocks until deployment completion. `nomad drain -monitor` has proven inadequate for fully understanding the drain process, so it might be a better idea to simply print the deployment's `status` output every time an update occurs.

A stretch goal would be detecting a tty and updating the output in place vs printing new output on each update, although that would also require a flag to disable (or proper detection of piping).

Another stretch goal would be adding a `-monitor` flag to `nomad job run` that would effectively switch from the current `job run` output to the new `deployment status -monitor` output after the normal `job run` output completed.

### Prior art

- `kubectl rollout status` - https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#deployment-status

**Comment by krishicks** [comment]
> simply print the deployment's status output every time an update occurs.

I agree with this, though a CLI user might want updates even when a change hasn't occurred just to show the process isn't just hanging. I implemented this status checking without going the extra step of writing something out even if the status doesn't change and it doesn't feel right.

I still think we should limit the API calls, but hide that fact behind some constant progress updates.

This is my implementation: https://github.com/hashicorp/cloud-sre/blob/master/cli/lib/remote/command_watch.sh

**Comment by krishicks** [comment]
I've added a `--watch` flag to Cloud's `deploy` command which includes output that mimics nomad's own output:

```
==> Monitoring evaluation "e7bef548"
    Evaluation triggered by job "cadence-worker"
    Evaluation within deployment: "502d4768"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "e7bef548" finished with status "complete"
==> Monitoring deployment "502d4768"
    Deployment updated: status is "running"
    Deployment updated: status is "running"
    Deployment updated: status is "running"
    Deployment updated: status is "running"
    Deployment updated: status is "running"
    Deployment updated: status is "running"
    Deployment updated: status is "successful"
```

**Comment by cricket007** [comment]
Is this exposed via the API at all?

**Comment by tgross** [comment]
Hi @cricket007! The existing `nomad node drain -monitor` works by polling the [`v1/node/:node_id`](https://nomadproject.io/api-docs/nodes/#read-node) endpoint (see [api/nodes.go#L216](https://github.com/hashicorp/nomad/blob/dc39c11d06c1744bab152343633119cb224ce39f/api/nodes.go#L216) for the implementation). Any deployment monitoring would likely work the same way.

**Comment by tcurdt** [comment]
Is there any chance to also see the allocations?

Either way: I would love to see this get released. Are there any plans for a release yet?

**Comment by tgross** [comment]
Hi @tcurdt! Unfortunately we don't yet have a public roadmap for Nomad, so I can't promise you when this will land. But it's definitely something we want to do.
